crates/volta-core/src/hook/mod.rs

```diff
@@ -68,27 +68,106 @@ pub struct ToolHooks<D: Distro> {
     pub phantom: PhantomData<D>,
 }
 
+impl<D: Distro> ToolHooks<D> {
+    /// Creates a merged struct, with "right" having precedence over "left".
+    fn merge(left: Self, right: Self) -> Self {
+        Self {
+            distro: right.distro.or(left.distro),
+            latest: right.latest.or(left.latest),
+            index: right.index.or(left.index),
+            phantom: PhantomData,
+        }
+    }
+}
+
 impl HookConfig {
-    /// Returns the current hooks, loaded from the filesystem.
+    /// Returns the current hooks, which are a merge between the user hooks and
+    /// the project hooks (if any).
     fn current() -> Fallible<Self> {
-        let path = user_hooks_file()?;
-        let src = touch(&path)
-            .and_then(|mut file| file.read_into_string())
-            .with_context(|_| ErrorDetails::ReadHooksError {
-                file: path.to_string_lossy().to_string(),
-            })?;
-        src.parse()
+        let maybe_project_config = Self::for_current_dir()?;
+        let user_config = Self::for_user()?;
+
+        Ok(if let Some(project_config) = maybe_project_config {
+            Self::merge(user_config, project_config)
+        } else {
+            user_config
+        })
+    }
+
+    /// Returns the per-project hooks for the current directory.
+    fn for_current_dir() -> Fallible<Option<Self>> {
+        Self::for_dir(&env::current_dir().with_context(|_| ErrorDetails::CurrentDirError)?)
+    }
+
+    /// Returns the per-project hooks for the specified directory.  If the
+    /// specified directory is not itself a project, its ancestors will be
+    /// searched.
+    fn for_dir(base_dir: &Path) -> Fallible<Option<Self>> {
+        match project_for_dir(&base_dir) {
+            Some(project_dir) => {
+                let path = project_dir.join(".volta").join("hooks.json");
+
+                if !path.is_file() {
+                    return Ok(None);
+                }
+
+                Self::from_file(&path, false).map(|hooks| Some(hooks))
+            }
+            None => Ok(None),
+        }
     }
-}
 
-impl FromStr for HookConfig {
-    type Err = VoltaError;
+    fn from_file(file_path: &Path, create: bool) -> Fallible<Self> {
```

Comment:
I'm a little wary of boolean parameters in methods, because it's difficult to tell from the call-site what that `true` or `false` value means. I think we can achieve this behavior by checking for an empty string instead of creating a file with `{}` as the contents:

```rust
fn from_file(file_path: &Path) -> Fallible<Self> {
    let src = touch(&file_path)
        .and_then(|mut file| file.read_into_string())
        .with_context(|_| ErrorDetails::ReadHooksError {
            file: file_path.to_string_lossy().to_string(),
        })?;

    if src.is_empty() {
        HookConfig {
            node: None,
            yarn: None,
            package: None,
            events: None,
        }
    } else {
        let serial: serial::HookConfig =
            serde_json::de::from_str(src).with_context(|_| ErrorDetails::ParseHooksError {
                file: file_path.to_string_lossy().to_string(),
            })?;
        serial.into_hook_config()
    }
}
```

Reply:
I'm not happy with the boolean argument, either, but `from_file` can't blindly call `touch`, because it's also used in the per-project hooks, where we **don't** want it to create a file when one doesn't already exist.

Really, I'd be happier not generating the per-user file, though that would be a departure from existing behavior.

Reply:
When I *have* to do something like that (it happens sometimes) I just create a standalone, two-value enum to make the invocation clear.

Reply:
It appears from lines 110-114 above, that the per-project hook part only calls `from_file` if the file already exists on the disk, so the use of `touch` shouldn't change anything, correct?

Reply:
I'd forgotten that!  Let me make these changes.